packages/react-async/src/component.tsx

```diff
@@ -1,135 +1,330 @@
-import * as React from 'react';
-import {LoadProps, DeferTiming} from '@shopify/async';
-import {Props as ComponentProps} from '@shopify/useful-types';
+import React, {
+  useEffect,
+  useRef,
+  useCallback,
+  ReactNode,
+  ComponentType,
+} from 'react';
 
-import {Async, AsyncPropsRuntime} from './Async';
+import {createResolver, ResolverOptions, DeferTiming} from '@shopify/async';
+import {OnIdle, useIdleCallback} from '@shopify/react-idle';
+import {IntersectionObserver} from '@shopify/react-intersection-observer';
+import {Hydrator, useHydrationManager} from '@shopify/react-hydrate';
 
-interface ConstantProps {
-  async?: AsyncPropsRuntime;
-}
+import {useAsync} from './hooks';
+import {AssetTiming, AsyncComponentType} from './types';
 
 interface Options<
-  Props,
-  PreloadProps = {},
-  PrefetchProps = {},
-  KeepFreshProps = {}
-> extends LoadProps<React.ComponentType<Props>> {
-  defer?: DeferTiming;
-  renderLoading?(): React.ReactNode;
-  renderPreload?(props?: PreloadProps): React.ReactNode;
-  renderPrefetch?(props?: PrefetchProps): React.ReactNode;
-  renderKeepFresh?(props?: KeepFreshProps): React.ReactNode;
-}
-
-export interface AsyncComponentType<
-  Props,
-  PreloadProps,
-  PrefetchProps,
-  KeepFreshProps
-> {
-  (props: Props & ConstantProps): React.ReactElement<
-    ComponentProps<typeof Async>
-  >;
-  Preload(props: PreloadProps & ConstantProps): React.ReactElement<{}>;
-  Prefetch(props: PrefetchProps & ConstantProps): React.ReactElement<{}>;
-  KeepFresh(props: KeepFreshProps & ConstantProps): React.ReactElement<{}>;
+  Props extends object,
+  PreloadOptions extends object = {},
+  PrefetchOptions extends object = {},
+  KeepFreshOptions extends object = {}
+> extends ResolverOptions<ComponentType<Props>> {
+  defer?: DeferTiming | ((props: Props) => boolean);
+  deferHydration?: DeferTiming | ((props: Props) => boolean);
+  displayName?: string;
+  renderLoading?(props: Props): ReactNode;
+  renderError?(error: Error): ReactNode;
+  usePreload?(props: PreloadOptions): () => void;
+  usePrefetch?(props: PrefetchOptions): () => void;
+  useKeepFresh?(props: KeepFreshOptions): () => void;
 }
 
 export function createAsyncComponent<
-  Props,
-  PreloadProps = {},
-  PrefetchProps = {},
-  KeepFreshProps = {}
+  Props extends object,
+  PreloadOptions extends object = {},
+  PrefetchOptions extends object = {},
+  KeepFreshOptions extends object = {}
 >({
   id,
   load,
   defer,
-  renderLoading,
-  renderPreload = noopRender,
-  renderPrefetch = noopRender,
-  renderKeepFresh = noopRender,
+  deferHydration,
+  displayName,
+  renderLoading = noopRender,
+  renderError = defaultRenderError,
+  usePreload: useCustomPreload = noopUse,
```

Comment:
Ah so they are named with `use` because they are actually invoked as hooks in here. 🤔

Can we add some like JSDoc to this function or something that specifies that they are directly used as hooks?

Reply:
Isn’t that what useX signals everywhere else?

Reply:
yeah, I would just weird when they are parameters to a method

Reply:
Yeah I agree with that, I'll add some JSDoc to it.